index.js

```diff
@@ -1,93 +1,138 @@
+    this._oErrorTags // eslint-disable-line
+  }
+
+  /**
+   * Set the extra info object for this error.
+   *
+   * @param {Object | null | undefined} info extra data to attach to the error
+   * @return {this}
+   */
+  withInfo(info) {
+    this.info = info
+    return this
   }
 
   /**
    * Wrap the given error, which caused this error.
    *
-   * @param {Error} cause
+   * @param {Error} cause the internal error that caused this error
    * @return {this}
    */
   withCause(cause) {
     this.cause = cause
-    if (this.message && cause.message) {
-      this.message += ': ' + cause.message
-    }
     return this
   }
-}
 
-/**
- * Return the `info` property from `error` and recursively merge the `info`
- * properties from the error's causes, if any.
- *
- * If a property is repeated, the first one in the cause chain wins.
- *
- * @param {?Error} error assumed not to have circular causes
- * @return {Object}
- */
-function getFullInfo(error) {
-  if (!error) return {}
-  const info = getFullInfo(error.cause)
-  if (typeof error.info === 'object') Object.assign(info, error.info)
-  return info
-}
+  /**
+   * Tag debugging information onto any error (whether an OError or not) and
+   * return it.
+   *
+   * @param {Error} error the error to tag
+   * @param {string} [message] message with which to tag `error`
+   * @param {Object} [info] extra data with wich to tag `error`
+   * @return {Error} the modified `error` argument
+   */
+  static tag(error, message, info) {
+    const oError = /** @type{OError} */ (error)
 
-/**
- * Return the `stack` property from `error` and recursively append the `stack`
- * properties from the error's causes, if any.
- *
- * @param {?Error} error assumed not to have circular causes
- * @return {string}
- */
-function getFullStack(error) {
-  if (!error) return ''
-  const causeStack = getFullStack(error.cause)
-  if (causeStack) return error.stack + '\ncaused by: ' + causeStack
-  return error.stack
+    if (!oError._oErrorTags) oError._oErrorTags = []
+
+    let tag
+    if (Error.captureStackTrace) {
+      // Hide this function in the stack trace, and avoid capturing it twice.
+      tag = /** @type TaggedError */ ({ name: 'TaggedError', message, info })
+      Error.captureStackTrace(tag, OError.tag)
+    } else {
+      tag = new TaggedError(message, info)
+    }
+
+    oError._oErrorTags.push(tag)
+
+    return error
+  }
+
+  /**
+   * The merged info from any `tag`s on the given error.
+   *
+   * If an info property is repeated, the last one wins.
+   *
+   * @param {Error | null | undefined} error any errror (may or may not be an `OError`)
+   * @return {Object}
+   */
+  static getFullInfo(error) {
+    const info = {}
+
+    if (!error) return info
+
+    const oError = /** @type{OError} */ (error)
+
+    if (typeof oError.info === 'object') Object.assign(info, oError.info)
+
+    if (oError._oErrorTags) {
+      for (const tag of oError._oErrorTags) {
+        Object.assign(info, tag.info)
```

Comment:
It seems that later info items will overwrite earlier ones with the same name. Is that OK? Is there any reason we'd want to preserve the whole set of `info` key-value pairs?

Reply:
It's a good question... this combined 'full info' object is what gets logged:
https://github.com/overleaf/logger-module/blob/a80466da8a51ddd75cd9113b55fdcdb7a31dca30/logging-manager.js#L16

I don't think there's anything in principle preventing us from instead logging out an array of `info` objects, but it would probably make it harder to write structured logging queries on the info if it were all wrapped up in the elements of an array (and it would be harder to read the logs). So, there's some risk that we clobber some of the info here, but I think it's probably worth taking that risk for the simpler output.

If this does become a problem, I think we could:
- change it to be 'first writer wins' rather than 'last writer wins', but I'm not sure there's a clear winner between those two strategies
- change the merge logic to be smarter --- if the key and value are the same, let it be, but if we get two values for the same key, either turn the value into an array or add a second key (`userId2` say).

That might be one for v3.1.

Reply:
Yup, I'm fine with leaving it as it is for now :+1: